File: src/main/java/com/danyazero/utils/ArrayType.java
package com.danyazero.utils;

import com.danyazero.model.PrimitiveType;
import com.danyazero.model.Type;
import org.objectweb.asm.MethodVisitor;

import static org.objectweb.asm.Opcodes.*;

public class ArrayType implements Type<Integer> {
    private final Type<?> child;

    public ArrayType(Type<?> type) {
        this.child = type;
    }

    public Type<?> getChild() {
        return child;
    }

    @Override
    public void postack(MethodVisitor mv, Integer value) {
        new IntegerType().postack(mv, value);

        if (child instanceof PrimitiveType<?> type) {
            System.out.println("New array.");
            mv.visitIntInsn(NEWARRAY, type.getType());
        } else {
            mv.visitInsn(ANEWARRAY);
        }
    }

    @Override
    public void store(MethodVisitor mv, short index) {
        mv.visitVarInsn(ASTORE, index);
    }

    @Override
    public void load(MethodVisitor mv, short index) {
        mv.visitVarInsn(ALOAD, index);
    }

    @Override
    public void yield(MethodVisitor mv) {
    }

    @Override
    public void aload(MethodVisitor mv) {
        throw new  RuntimeException("Not implemented");
    }

    @Override
    public void astore(MethodVisitor mv) {
        throw new  RuntimeException("Not implemented");
    }


    private void loadElement(MethodVisitor mv, int index) {
        if (child instanceof PrimitiveType<?> type) {
            new IntegerType().postack(mv, index);
            type.aload(mv);
        }

        throw new RuntimeException("This type is not acceptable.");
    }

    @Override
    public short getSize() {
        return 1;
    }
}

File: src/main/kotlin/type/ArrayType.kt
package com.danyazero.type

import com.danyazero.model.PrimitiveType
import com.danyazero.model.ReferenceType
import com.danyazero.model.Type
import org.objectweb.asm.Label
import org.objectweb.asm.MethodVisitor
import org.objectweb.asm.Opcodes

class ArrayType(
    val child: Type<*>,
    val length: Int? = null
) : Type<Int> {

    override fun postack(mv: MethodVisitor, value: Int) {
        IntegerType().postack(mv, value)

        if (child is PrimitiveType<*>) {
            mv.visitIntInsn(Opcodes.NEWARRAY, child.getType())
        } else if (child is ReferenceType){
            mv.visitTypeInsn(Opcodes.ANEWARRAY, "java/lang/String")
        }
    }

    fun length(mv: MethodVisitor) {
        mv.visitInsn(Opcodes.ARRAYLENGTH)
    }

    override fun store(mv: MethodVisitor, index: Short) {
        mv.visitVarInsn(Opcodes.ASTORE, index.toInt())
    }

    override fun load(mv: MethodVisitor, index: Short) {
        mv.visitVarInsn(Opcodes.ALOAD, index.toInt())
    }

    override fun yield(mv: MethodVisitor) {
    }

    override fun aload(mv: MethodVisitor) {
        throw RuntimeException("Not implemented")
    }

    override fun astore(mv: MethodVisitor) {
        throw RuntimeException("Not implemented")
    }

    override fun equal(mv: MethodVisitor, jumpTarget: Label) {
        //TODO ArrayType equals operation
    }


    fun loadElement(mv: MethodVisitor, index: Int) {
        if (child is PrimitiveType<*>) {
            IntegerType().postack(mv, index)
            child.aload(mv)
        }

        throw RuntimeException("This type is not acceptable.")
    }

    override fun getSize(): Short {
        return 1
    }

    override fun toString(): String {
        return "ArrayType(child=$child)"
    }
}